pkg/pmem-csi-driver/rescheduler.go

```diff
@@ -0,0 +1,215 @@
+	if opts.SelectedNode != nil {
+		err = &controller.IgnoredError{
+			Reason: fmt.Sprintf("not responsible for provisioning of PVC %s/%s because it will be handled by the PMEM-CSI driver on node %q",
+				opts.PVC.Namespace, opts.PVC.Name, opts.SelectedNode.Name),
+		}
+	} else {
+		err = &controller.IgnoredError{
+			Reason: fmt.Sprintf("not responsible for provisioning of PVC %s/%s because it is not assigned to a node",
+				opts.PVC.Namespace, opts.PVC.Name),
+		}
+	}
+	return nil, controller.ProvisioningNoChange, err
+}
+
+func (pcp *pmemCSIProvisioner) Delete(context.Context, *v1.PersistentVolume) error {
+	return errors.New("not implemented")
+}
+
+func (pcp *pmemCSIProvisioner) shouldReschedule(ctx context.Context, pvc *v1.PersistentVolumeClaim, node *v1.Node) (bool, error) {
+	l := logger.Get(ctx).WithName("ShouldReschedulePVC").WithValues("pvc", klog.KObj(pvc))
+	if node != nil {
+		l = l.WithValues("node", klog.KObj(node))
+	}
+
+	// "node" might be nil. Check the label directly.
+	var selectedNode string
+	if pvc.Annotations != nil {
+		selectedNode = pvc.Annotations[annSelectedNode]
+	}
+	if selectedNode == "" {
+		// No need to reschedule.
+		l.V(5).Info("no need to reschedule, no selected node")
+		return false, nil
+	}
+
+	// We have to be absolutely certain that the PVC is not going
+	// to be handled on the node. If we remove the annotation
+	// while a driver node instance starts to provisision it,
+	// volumes may leak.
+	//
+	// Therefore we check both the labels on the node ("Should a
+	// PMEM-CSI driver run here?") and the CSINode object ("Does a
+	// PMEM-CSI driver (still) run here?").
+	//
+	// The node check is expensive. We either would have to watch
+	// all nodes (which is expensive) or do a GET per check (also
+	// expensive). To mitigate this, the node is only checked when
+	// called through Provision() with a Node object already
+	// retrieved by the lib.
+	//
+	// When the scheduler extensions work, we should rarely get to
+	// Provision() because typically PVCs get assigned to nodes
+	// with PMEM-CSI and thus the CSINode check already bails out
+	// of ShouldProvision.
+	//
+	// Only when the extensions are off, then Provision() may get
+	// called more often. Such a cluster setup should better be
+	// avoided.
+	driverIsRunning := true
+	csiNode, err := pcp.csiNodeLister.Get(selectedNode)
+	switch {
+	case err == nil:
+		driverIsRunning = hasDriver(csiNode, pcp.driverName)
+	case apierrs.IsNotFound(err):
+		driverIsRunning = false
+	default:
+		return false, fmt.Errorf("retrieve CSINode %s: %v", selectedNode, err)
+	}
+	if node == nil {
+		// Decide only based on CSINode.
+		reschedule := !driverIsRunning
+		l.V(3).Info("result", "reschedule", reschedule, "driverIsRunning", driverIsRunning)
+		return reschedule, nil
+	}
+
+	driverMightRun := true
+	for key, value := range node.Labels {
+		if node.Labels[key] == value {
+			driverMightRun = false
+		}
+	}
+
+	reschedule := !driverMightRun && !driverIsRunning
```

Comment:
Isn't it that we should make sure that the selected node has enough PMEM capacity instead of just checking if the driver is running on this node?

Reply:
If the driver is running, then it will do that check itself.

Checking capacity here is redundant and racy (we check capacity, fail, removed selected node, while sometime later the driver processes the PVC and creates a volume although it is no longer assigned to the node).

Reply:
In other words, the rescheduler only may handle the case where the driver definitely isn't running.